src/Monarch/Document.ts

```diff
@@ -7,79 +7,93 @@ import { asap } from 'asap/browser-asap'
+    const outputHandlers = OutputHandlersList.mkNil(dispatchMessage)
     const scheduler = mkScheduler()
     const environment = { scheduler, dispatchDiffWork, finishDiffWork }
 
     let model = initialModel
     let commitedVirtualDomTree = initialVirtualDomTree
     let diffWork: DiffWork<any, any> | undefined = undefined
+    let patchTree: PatchTree | undefined = undefined
 
-    let requestedAsyncRenderId: number | undefined = undefined
-    let requestedAsyncPerformDiffWorkId: number | undefined = undefined
+    let hasRequestedAsyncRendering = false
+    let hasRequestedAsyncPerformDiffWork = false
+    let hasRequestedAsyncCommitting = false
 
     function dispatchMessage(message: message): void {
         const previousModel = model
-        const nextModel = update(message)(model)
+        const nextModel = update(message)(previousModel)
 
         if (previousModel === nextModel) return
 
         model = nextModel
 
-        if (!requestedAsyncRenderId) {
-            requestedAsyncRenderId = requestAsap(render)
+        if (!hasRequestedAsyncRendering) {
+            hasRequestedAsyncRendering = requestAsap(render)
         }
     }
 
     function render() {
         const nextVirtualDomTree = view(model)
-        const initialDiffWork = mkDiffWork(commitedVirtualDomTree)(nextVirtualDomTree)
+        const initialDiffWork = mkDiffWork(commitedVirtualDomTree, nextVirtualDomTree)
 
-        requestedAsyncRenderId = undefined
+        hasRequestedAsyncRendering = false
 
         dispatchDiffWork(initialDiffWork)
     }
 
     function dispatchDiffWork(nextDiffWork: DiffWork<any, any>): void {
         diffWork = nextDiffWork
 
-        if (!requestedAsyncPerformDiffWorkId) {
-            requestedAsyncPerformDiffWorkId = window.setImmediate(() => {
-                unsafe_uncurried_performDiffWork(diffWork!, environment)
-            })
+        if (!hasRequestedAsyncPerformDiffWork) {
+            hasRequestedAsyncPerformDiffWork = true
+
+            window.setImmediate(performDiffWork)
         }
     }
 
-    function finishDiffWork({ rootVNode, rootPathTree }: FinishDiffWorkSpec<message>): void {
-        commitedVirtualDomTree = rootVNode
-        requestedAsyncPerformDiffWorkId = undefined
+    function performDiffWork() {
+        unsafe_uncurried_performDiffWork(diffWork!, environment)
+    }
 
-        requestAnimationFrame(() => {
-            unsafe_uncurried_applyPatchTree(container, rootPathTree)
-        })
+    function finishDiffWork({ rootVNode, rootPatchTree }: FinishDiffWorkSpec<message>): void {
+        patchTree = rootPatchTree
+        hasRequestedAsyncPerformDiffWork = false
+
+        if (!hasRequestedAsyncCommitting) {
+            hasRequestedAsyncCommitting = true
+
+            requestAnimationFrame(() => {
+                unsafe_uncurried_applyPatchTree(container, patchTree!)
+                commitedVirtualDomTree = rootVNode
+                hasRequestedAsyncCommitting = false
+            })
```

Comment:
This function can move to a higher binding level. Its allocation on every call of `finishDiffWork` is redundant.

Reply:
How can I get  `rootVNode` local variable in that function? with new state for that?

Reply:
We have discussed this in the DM but just for possible future readers: we should store the whole `FinishDiffWorkSpec` as the state instead of `PatchTree` only.

Reply:
Also, I have suggested `DiffWorkResult` name as a more understandable alternative to `FinishDiffWorkSpec`.